src/coins.cpp

```diff
@@ -1102,3 +1171,136 @@ CCoinsModifier::~CCoinsModifier()
         cache.cachedCoinsUsage += it->second.coins.DynamicMemoryUsage();
     }
 }
+
+void SubtreeCache::clear() {
+    initialized = false;
+    parentLatestSubtree = std::nullopt;
+    newSubtrees.clear();
+}
+
+std::optional<libzcash::LatestSubtree> SubtreeCache::GetLatestSubtree(CCoinsView *parentView) {
+    if (!initialized) {
+        parentLatestSubtree = parentView->GetLatestSubtree(type);
+        initialized = true;
+    }
+
+    if (newSubtrees.size() > 0) {
+        // The latest subtree is in our cache.
+
+        libzcash::SubtreeIndex index;
+        if (parentLatestSubtree.has_value()) {
+            // The best subtree index is newSubtrees.size() larger than
+            // our parent view's subtree index.
+            index = parentLatestSubtree.value().index + newSubtrees.size();
+        } else {
+            // The parent view has no subtrees
+            index = newSubtrees.size() - 1;
+        }
+
+        auto lastSubtree = newSubtrees.back();
+        return libzcash::LatestSubtree(index, lastSubtree.root, lastSubtree.nHeight);
+    } else {
+        return parentLatestSubtree;
+    }
+}
+
+std::optional<libzcash::SubtreeData> SubtreeCache::GetSubtreeData(CCoinsView *parentView, libzcash::SubtreeIndex index) {
+    auto latestSubtree = GetLatestSubtree(parentView);
+
+    if (!latestSubtree.has_value() || latestSubtree.value().index < index) {
+        // This subtree isn't complete in our local view
+        return std::nullopt;
+    }
+
+    if (parentLatestSubtree.has_value() && (index <= parentLatestSubtree.value().index)) {
+        // This subtree is complete, but it's not in the local cache
+        return parentView->GetSubtreeData(type, index);
+    } else {
+        if (parentLatestSubtree.has_value()) {
+            // Get the index into our local `newSubtrees` that the subtree should
+            // be located.
+            auto localIndex = index - (parentLatestSubtree.value().index + 1);
+            assert(newSubtrees.size() > localIndex);
```

Comment:
Why is this an assertion, rather than
```suggestion
            if (localIndex >= newSubtrees.size()) {
                return std::nullopt;
            }
```

Reply:
It should never be the case that `localIndex` is out of bounds here; that should be caught in the earlier bounds check of the function.

Reply:
Got it, I had misread that check.